File: samples/dynamicEntityDemoClient/src/uk/ac/soton/itinnovation/experimedia/arch/ecc/samples/dynamicEntityDemoClient/ECCClientController.java
/////////////////////////////////////////////////////////////////////////
//
// © University of Southampton IT Innovation Centre, 2012
//
// Copyright in this software belongs to University of Southampton
// IT Innovation Centre of Gamma House, Enterprise Road, 
// Chilworth Science Park, Southampton, SO16 7NS, UK.
//
// This software may not be used, sold, licensed, transferred, copied
// or reproduced in whole or in part in any manner or form or in or
// on any media by any person other than in accordance with the terms
// of the Licence Agreement supplied with the software, or otherwise
// without the prior written consent of the copyright owners.
//
// This software is distributed WITHOUT ANY WARRANTY, without even the
// implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
// PURPOSE, except where stated in the Licence Agreement supplied with
// the software.
//
//      Created By :            Simon Crowle
//      Created Date :          15-Aug-2012
//      Created for Project :   EXPERIMEDIA
//
/////////////////////////////////////////////////////////////////////////

package uk.ac.soton.itinnovation.experimedia.arch.ecc.samples.dynamicEntityDemoClient;

import uk.ac.soton.itinnovation.experimedia.arch.ecc.samples.shared.*;
import uk.ac.soton.itinnovation.experimedia.arch.ecc.amqpAPI.impl.amqp.*;

import uk.ac.soton.itinnovation.experimedia.arch.ecc.common.logging.spec.*;
import uk.ac.soton.itinnovation.experimedia.arch.ecc.common.loggin.impl.Log4JImpl;

import uk.ac.soton.itinnovation.experimedia.arch.ecc.common.dataModel.metrics.*;
import uk.ac.soton.itinnovation.experimedia.arch.ecc.common.dataModel.monitor.*;
import uk.ac.soton.itinnovation.experimedia.arch.ecc.common.dataModel.experiment.Experiment;

import java.util.*;
import javax.swing.JOptionPane;




public class ECCClientController implements EMIAdapterListener,
                                            ECCClientViewListener,
                                            ECCNewEntityViewListener
{
    private final IECCLogger clientLogger;

    private AMQPBasicChannel   amqpChannel;
    private EMInterfaceAdapter emiAdapter;
    private ECCClientView      clientView;
    private String             clientName;

    private MetricGenerator metricGenerator;
    private MetricGroup     metricGroup;
 


    public ECCClientController()
    {
        // Configure logging system
        Logger.setLoggerImpl( new Log4JImpl() );
        clientLogger = Logger.getLogger( ECCClientController.class );
      
        // Only one metric generator and metric group
        metricGenerator = new MetricGenerator();
        metricGenerator.setName( "Demo metric generator" );
        
        // Only one metric group
        metricGroup = MetricHelper.createMetricGroup( "Example group", "Group to add new metrics in", metricGenerator );
        
    }

    public void start( String rabbitServerIP,
                       UUID expMonitorID,
                       UUID clientID ) throws Exception
    {
        if ( rabbitServerIP != null &&
             expMonitorID   != null &&
             clientID       != null )
        {
            clientLogger.info( "Trying to connect to Rabbit server on " + rabbitServerIP );

            // Create connection to Rabbit server ------------------------------------
            AMQPConnectionFactory amqpFactory = new AMQPConnectionFactory();
            amqpFactory.setAMQPHostIPAddress( rabbitServerIP );
            try
            {
                amqpFactory.connectToAMQPHost();
                amqpChannel = amqpFactory.createNewChannel();
            }
            catch (Exception e ) 
            {
                clientLogger.error( "Could not connect to Rabbit server" );
                throw e; 
            }

            // Set up a simple view --------------------------------------------------
            Date date = new Date();
            clientName = date.toString();
            clientView = new ECCClientView( clientName, this, this );
            clientView.setVisible( true );

            // Create EM interface adapter, listen to it...
            emiAdapter = new EMInterfaceAdapter( this );

            // ... and try registering with the EM.

            try { emiAdapter.registerWithEM( clientName,
                                            amqpChannel, 
                                            expMonitorID, clientID ); }
            catch ( Exception e ) 
            {
                clientLogger.error( "Could not attempt registration with EM" );
                throw e; 
            }
        }
    }

    // EMIAdapterListener --------------------------------------------------------
    @Override
    public void onEMConnectionResult( boolean connected, Experiment expInfo )
    {
        if ( connected )
        {
          clientView.setStatus( "Connected to EM" );
          clientView.addLogMessage( "Linked to experiment: " + expInfo.getName() );
        }
        else
          clientView.setStatus( "Refused connection to EM" );
    }

    @Override
    public void onEMDeregistration( String reason )
    {
        clientView.addLogMessage( "Got disconnected from EM: " + reason );
        
        try
        { emiAdapter.disconnectFromEM(); }
        catch ( Exception e )
        { clientLogger.error( "Had problems disconnecting from EM: " + e.getMessage() ); }
        
        // Apologise to the user
        JOptionPane.showMessageDialog( null, "ECC disconnected this client: " + reason );
        
        clientView.dispose();
        System.exit(0);
    }
    
    @Override
    public void onDescribeSupportedPhases( EnumSet<EMPhase> phasesOUT )
    {
        // We're going to support all phases (although we won't do much in some of them)
        // ... we MUST support the discovery phase by default, but don't need to include
        phasesOUT.add( EMPhase.eEMSetUpMetricGenerators );
        phasesOUT.add( EMPhase.eEMLiveMonitoring );
        phasesOUT.add( EMPhase.eEMPostMonitoringReport );
        phasesOUT.add( EMPhase.eEMTearDown );
    }
    
    @Override
    public void onDescribePushPullBehaviours( Boolean[] pushPullOUT )
    {
        // We're going to support both push and pull
        pushPullOUT[0] = true;
        pushPullOUT[1] = true;
    }
    
    @Override
    public void onPopulateMetricGeneratorInfo()
    {
       HashSet<MetricGenerator> mgSet = new HashSet<MetricGenerator>();
       mgSet.add( metricGenerator );
       
       emiAdapter.sendMetricGenerators( mgSet );
   }

    @Override
    public void onDiscoveryTimeOut()
    { clientView.addLogMessage( "Got discovery time-out message" ); }
    
    @Override
    public void onSetupMetricGenerator( UUID genID, Boolean[] resultOUT )
    {
        clientView.setStatus( "Setting up generators" );

        // Just signal that the metric generator is ready
        resultOUT[0] = true;

        clientView.addLogMessage( "Completed generator set-up" );
    }
    
    @Override
    public void onSetupTimeOut( UUID metricGeneratorID )
    { clientView.addLogMessage( "Got set-up time-out message" ); }
    
    @Override
    public void onLiveMonitoringStarted()
    {
        clientView.addLogMessage( "ECC has started Live Monitoring process" );
    }

    @Override
    public void onStartPushingMetricData()
    {
        // Allow the human user to manually push some data
        clientView.addLogMessage( "Enabling metric push" );
        clientView.enablePush( true );
    }

    @Override
    public void onPushReportReceived( UUID reportID )
    {   
        // Got the last push, so allow another manual push
        clientView.enablePush( true );
    }
    
    @Override
    public void onPullReportReceived( UUID reportID )
    {

    }
    
    @Override
    public void onPullMetricTimeOut( UUID measurementSetID )
    { clientView.addLogMessage( "Got live pull time-out message" ); }

    @Override
    public void onStopPushingMetricData()
    {
        // Stop manual pushing of data by the human user
        clientView.addLogMessage( "Disabling metric push" );
        clientView.enablePush( false );
    }

    @Override
    /*
    * Note that 'reportOut' is an OUT parameter provided by the adapter
    */
    public void onPullMetric( UUID measurementSetID, Report reportOut)
    {
        // Create an empty instance of our measurement set
        MeasurementSet targetSet = MetricHelper.getMeasurementSet( metricGenerator, measurementSetID );

        // Create a copy of the measurement set, but with no measurements in
        Report newReport = MetricHelper.createEmptyMeasurementReport( targetSet );
        
        // Generate a simulated measurement using a random number
        int ran = 1 +(int)(Math.random()*20);
        String mes = ""+ ran;
        Measurement m = new Measurement( mes );
        
        // Add measurement
        newReport.getMeasurementSet().addMeasurement( m );
        newReport.setNumberOfMeasurements( 1 );
       
        // Copy report into OUT parameter
        reportOut.copyReport( newReport, true );
    }

    @Override
    public void onPullingStopped()
    {
        clientView.addLogMessage( "ECC has stopped pulling" );
    }
    
    @Override
    /*
    * Note that the summaryOUT parameter is an OUT parameter supplied by the
    * adapter
    */
    public void onPopulateSummaryReport( EMPostReportSummary summaryOUT )
    {
        // We've only got one MeasurementSet so we'll create a demo summary report
        // and just two measurements.. so we'll use these to create a demo summary

        // If we don't have any measurements, make some up!
//        if ( firstMeasurement   ==  null ) snapshotMeasurement();
//        if ( currentMeasurement ==  null ) snapshotMeasurement();
//
//        // Create a new report for this summary
//        Report report = new Report();
//        report.setReportDate( new Date() );
//        report.setFromDate( firstMeasurement.getTimeStamp() );
//        report.setToDate( currentMeasurement.getTimeStamp() );
//        report.setNumberOfMeasurements( 2 );
//
//        // We've only got one of each...
//        MetricGenerator mGen = metricGenerators.values().iterator().next();
//        MetricGroup mGroup = mGen.getMetricGroups().iterator().next();
//        MeasurementSet mSet = mGroup.getMeasurementSets().iterator().next();
//
//        // Add our one MeasurementSet data to the report and add that to the summary report
//        report.setMeasurementSet( mSet );
//        summaryOUT.addReport( report );
    }

    @Override
    public void onPopulateDataBatch( EMDataBatch batchOUT )
    {
        // We've only stored the first and the last measurements of a single
        // MeasurementSet, so just send that
//        MeasurementSet ms = createMeasurementSetEmptySample();
//        ms.addMeasurement( firstMeasurement );
//        ms.addMeasurement( currentMeasurement );
//        
//        Report batchRep = new Report();
//        batchRep.setFromDate( firstMeasurement.getTimeStamp() );
//        batchRep.setToDate( currentMeasurement.getTimeStamp() );
//        batchRep.setMeasurementSet( ms );
//        batchRep.setNumberOfMeasurements( 2 );
//        
//        batchOUT.setBatchReport( batchRep );
    }
    
    @Override
    public void onReportBatchTimeOut( UUID batchID )
    { clientView.addLogMessage( "Got post-report time-out message" ); }

    @Override
    public void onGetTearDownResult( Boolean[] resultOUT )
    {
        clientView.setStatus( "Tearing down" );
        clientView.addLogMessage( "Tearing down metric generators" );

        // Signal we've successfully torn-down
        resultOUT[0] = true;
    }

    // EMClientViewListener ------------------------------------------------------
    @Override
    public void onPushDataClicked()
    {
        // Create an empty instance of our measurement set
//        MeasurementSet sampleSet = createMeasurementSetEmptySample();
//
//        // Take a current measurement
//        snapshotMeasurement();
//        sampleSet.addMeasurement( currentMeasurement );
//
//        // Set up report (has only a single measure)
//        Date date           = new Date();
//        Report sampleReport = new Report();
//
//        sampleReport.setMeasurementSet( sampleSet );
//        sampleReport.setReportDate(date );
//        sampleReport.setFromDate( date );
//        sampleReport.setToDate( date );
//        sampleReport.setMeasurementSet( sampleSet );
//        sampleReport.setNumberOfMeasurements( 1 );
//
//        // ... and store for confirmation of push, then report!
//        pendingPushReports.put( sampleReport.getUUID(), sampleReport );
//        emiAdapter.pushMetric( sampleReport );
    }
    
    @Override
    public void onEntityMetricCollectionEnabled( UUID senderID, UUID entityID, boolean enabled )
    {
    
    }
    
    @Override
    public void onClientViewClosed()
    {
        // Need to notify that we're leaving...
        try { emiAdapter.disconnectFromEM(); }
        catch ( Exception e )
        { clientLogger.error( "Could not cleanly disconnect from EM:\n" + e.getMessage() ); }
    }
    
    @Override
    public void onTearDownTimeOut()
    { clientView.addLogMessage( "Got tear-down time-out message" ); }

    
    // ECCNewEntityViewListener events -----------------------------------------
    @Override
    public void onNewEntityInfoEntered(String entityName, String attName, String attDesc )
    {

        // Create a new entity to be observed (this Java VM)
        Entity entityBeingObserved = new Entity();
        entityBeingObserved.setName( entityName );

        // Create an attribute to observe
        Attribute entityAttribute = MetricHelper.createAttribute(attName, attDesc, entityBeingObserved);

        // Link entity with metric generator
        metricGenerator.addEntity( entityBeingObserved );
        
        // ... a single MeasurementSet (representing the measures for the attibute)
        MetricHelper.createMeasurementSet( entityAttribute, 
                                           MetricType.RATIO, 
                                           new Unit("Request count"), 
                                           metricGroup );
        
        clientView.addLogMessage( "Created new entity: " + entityBeingObserved.getName() );        
    }
    
    
    // Private method ------------------------------------------------------------
//    private MeasurementSet createMeasurementSetEmptySample()
//    {
//        // Get our only metric generator
//        MetricGenerator metGen = metricGenerators.values().iterator().next();
//        metGen.getMetricGroups().iterator().next();
//
//        // Get our only metric group
//        MetricGroup mg = metGen.getMetricGroups().iterator().next();
//        MeasurementSet currentMS = mg.getMeasurementSets().iterator().next();
//
//        return new MeasurementSet( currentMS, false );
//    }

//    private void snapshotMeasurement()
//    {
//        // Just take a very rough measurement
//        Runtime rt = Runtime.getRuntime();
//        String memVal = Long.toString( rt.totalMemory() - rt.freeMemory() );
//
//        // Get the (single) measurement set for this snapshot (just to get the ID actually)
//        MeasurementSet snapshotMS = createMeasurementSetEmptySample();
//
//        // Update the latest measurement
//        currentMeasurement = new Measurement();
//        currentMeasurement.setMeasurementSetUUID( snapshotMS.getID() );
//        currentMeasurement.setTimeStamp( new Date() );
//        currentMeasurement.setValue( memVal );
//
//        // Store this if it is the first ever measurement
//        if ( firstMeasurement == null ) firstMeasurement = currentMeasurement;
//
//        clientView.addLogMessage( "Memory measurement (bytes): " + memVal );
//    }
}

File: samples/basicDotNetClient/ECCClientController.cs
/////////////////////////////////////////////////////////////////////////
//
// © University of Southampton IT Innovation Centre, 2012
//
// Copyright in this software belongs to University of Southampton
// IT Innovation Centre of Gamma House, Enterprise Road, 
// Chilworth Science Park, Southampton, SO16 7NS, UK.
//
// This software may not be used, sold, licensed, transferred, copied
// or reproduced in whole or in part in any manner or form or in or
// on any media by any person other than in accordance with the terms
// of the Licence Agreement supplied with the software, or otherwise
// without the prior written consent of the copyright owners.
//
// This software is distributed WITHOUT ANY WARRANTY, without even the
// implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
// PURPOSE, except where stated in the Licence Agreement supplied with
// the software.
//
//      Created By :            Simon Crowle
//      Created Date :          10-Apr-2013
//      Created for Project :   EXPERIMEDIA
//
/////////////////////////////////////////////////////////////////////////

using uk.ac.soton.itinnovation.experimedia.arch.ecc.samples.shared;
using uk.ac.soton.itinnovation.experimedia.arch.ecc.amqpAPI.impl.amqp;
using uk.ac.soton.itinnovation.experimedia.arch.ecc.common.dataModel.metrics;
using uk.ac.soton.itinnovation.experimedia.arch.ecc.common.dataModel.experiment;
using uk.ac.soton.itinnovation.experimedia.arch.ecc.common.dataModel.monitor;

using System.Collections.Generic;
using System.Diagnostics;




namespace uk.ac.soton.itinnovation.experimedia.arch.ecc.samples.basicDotNetClient
{

public class ECCClientController : EMIAdapterListener
{
    private log4net.ILog clientLogger;

    private AMQPBasicChannel   amqpChannel;
    private EMInterfaceAdapter emiAdapter;
    private string             clientName;

    delegate Measurement ITakeMeasurement();
    private  Dictionary<System.Guid, ITakeMeasurement> samplers;

    public ECCClientController()
    {
        // Configure logging system
        clientLogger = log4net.LogManager.GetLogger(System.Reflection.MethodBase.GetCurrentMethod().DeclaringType);
    }
    
    public void start( string rabbitServerIP,
                       string username,
                       string password,
                       System.Guid expMonitorID,
                       System.Guid clientID )
    {
        if ( rabbitServerIP != null &&
             expMonitorID   != null &&
             clientID       != null )
        {
            clientLogger.Info( "Trying to connect to Rabbit server on " + rabbitServerIP );

            // Create connection to Rabbit server -------------------------------
            AMQPConnectionFactory amqpFactory = new AMQPConnectionFactory();
            amqpFactory.setAMQPHostIPAddress( rabbitServerIP );
            try
            {
                if (username != null && password != null)
                    amqpFactory.setLoginDetails(username, password);

                amqpFactory.connectToAMQPHost();
                amqpChannel = amqpFactory.createNewChannel();
            }
            catch (System.Exception e ) 
            {
                clientLogger.Error( "Could not connect to Rabbit server" );
                throw e; 
            }

            // Create a simple name
            System.DateTime date = System.DateTime.Now;
            clientName = "C# Client (" + date.ToString() + ")";

            // Create EM interface adapter, listen to it...
            emiAdapter = new EMInterfaceAdapter( this );

            // ... and try registering with the ECC -----------------------------
            try { emiAdapter.registerWithEM( clientName,
                                             amqpChannel, 
                                             expMonitorID, clientID ); }
            catch ( System.Exception e ) 
            {
                clientLogger.Error( "Could not attempt registration with ECC" );
                throw e; 
            }
        }
    }

    public void stop()
    {
        if (emiAdapter != null) emiAdapter.disconnectFromEM();
    }

    // EMIAdapterListener --------------------------------------------------------
    public void onEMConnectionResult( bool connected, Experiment expInfo )
    {
        if ( connected )
        {
          clientLogger.Info( "Connected to EM" );
          clientLogger.Info( "Linked to experiment: " + expInfo.name );
        }
        else
          clientLogger.Info( "Refused connection to EM" );
    }

    public void onEMDeregistration( string reason )
    {
        clientLogger.Info( "Got disconnected from EM: " + reason );
        
        try
        { emiAdapter.disconnectFromEM(); }
        catch ( System.Exception e )
        { clientLogger.Error( "Had problems disconnecting from EM: " + e.Message ); }
        
        // Apologise to the user
        clientLogger.Info( "ECC disconnected this client: " + reason );

        System.Environment.Exit(0);
    }
    
    public void onDescribeSupportedPhases( HashSet<EMPhase> phasesREF )
    {
        // We are going to just support live monitoring
        // ... we MUST support the discovery phase by default, but don't need to include

        clientLogger.Info("Specifying phase compatibility (Live monitoring only)");
        phasesREF.Add( EMPhase.eEMLiveMonitoring );
    }
    
    public void onDescribePushPullBehaviours( ref bool[] pushPullREF )
    {
        // We're going to support both push and pull
        clientLogger.Info("Specifying only pull behaviour");

        pushPullREF[0] = false;  // In this demo we are just going to be pull
        pushPullREF[1] = true;   // ECC will pull metrics from this client
    }
    
    public void onPopulateMetricGeneratorInfo()
    {
        clientLogger.Info("Creating metric generator info...");

        // Describe what we are going measure
        Entity app = new Entity( System.Guid.NewGuid(),
                                 "Client app",
                                 "The thing we are observing");

        Attribute phyRAMUage = MetricHelper.createAttribute( "Physical memory usage",
                                                             "Physical memory used by the app",
                                                             app );

        Attribute virtualMEMUsage = MetricHelper.createAttribute( "Virtual memory usage",
                                                                  "Virtual memory used by the app",
                                                                  app );

        // Create a data structure in which to put measurements
        MetricGenerator mGen = new MetricGenerator( System.Guid.NewGuid(),
                                                    "Client metric generator",
                                                    "The only metric generator for this client" );
        // Associate an entity with the generator
        mGen.addEntity(app);

        // Create a group of metric relating to memory
        MetricGroup group = MetricHelper.createMetricGroup( "Memory metrics",
                                                            "Memory metrics relating to the app",
                                                            mGen );

        // Create measurment sets for our two metrics (and use Guids to point to sampling methods)
        System.Guid physMem_ID = MetricHelper.createMeasurementSet( phyRAMUage,
                                                                    MetricType.RATIO,
                                                                    new Unit("bytes"),
                                                                    group )
                                                                    .msetID;

        System.Guid virtualMem_ID = MetricHelper.createMeasurementSet( virtualMEMUsage,
                                                                       MetricType.RATIO,
                                                                       new Unit("bytes"),
                                                                       group )
                                                                       .msetID;

        // Point to sampling delegates for each metric
        samplers = new Dictionary<System.Guid, ITakeMeasurement>();
        samplers.Add(physMem_ID,    new ITakeMeasurement(samplePhysicalMEM));
        samplers.Add(virtualMem_ID, new ITakeMeasurement(sampleVirtualMEM));

        // Describe what we have created on the console
        clientLogger.Info(MetricHelper.describeGenerator(mGen));

        // Wrap up metric generators and send to the ECC
        HashSet<MetricGenerator> mgSet = new HashSet<MetricGenerator>();
        mgSet.Add(mGen);

        emiAdapter.setMetricGenerators(mgSet);        
    }

    public void onDiscoveryTimeOut()
    { clientLogger.Info( "Got discovery time-out message" ); }
    
    public void onSetupMetricGenerator( System.Guid genID, ref bool[] resultREF )
    { /*Not implemented in this demo */ }
    
    public void onSetupTimeOut( System.Guid metricGeneratorID )
    { /*Not implemented in this demo */ }
    
    public void onLiveMonitoringStarted()
    {
        clientLogger.Info( "ECC has started Live Monitoring process" );
    }

    public void onStartPushingMetricData()
    {
        // Allow the human user to manually push some data
        clientLogger.Info( "Enabling metric push" );
    }

    public void onPushReportReceived( System.Guid reportID )
    { /*Not implemented in this demo */ }
    
    public void onPullReportReceived( System.Guid reportID )
    {
    }

    public void onPullMetricTimeOut( System.Guid measurementSetID )
    { clientLogger.Info( "Got live pull time-out message" ); }

    public void onStopPushingMetricData()
    {
        clientLogger.Info( "Disabling metric push" );
    }

    /*
    * Note that 'reportREF' is an out-style parameter provided by the adapter
    */
    public void onPullMetric( System.Guid measurementSetID, Report reportREF )
    {
        if (measurementSetID != null)
        {
            ITakeMeasurement sampler = samplers[measurementSetID];

            if (sampler != null)
            {
                Measurement m = sampler();
                reportREF.measurementSet.addMeasurement(m);
                reportREF.numberOfMeasurements = 1;

                clientLogger.Info("Send pull measurement for " + measurementSetID.ToString());
            }
            else clientLogger.Info("Got a pull, but could not find sampler for measurement requested");
        }
        else clientLogger.Info("Got a pull without a valid guid.");
    }

    public void onPullingStopped()
    {
        clientLogger.Info( "ECC has stopped pulling" );
    }
    
    /*
    * Note that the summaryOUT parameter is an OUT parameter supplied by the
    * adapter
    */
    public void onPopulateSummaryReport( EMPostReportSummary summaryREF )
    { /* Not implemented in this demo */ }

    public void onPopulateDataBatch( EMDataBatch batchREF )
    { /* Not implemented in this demo */ }
    
    public void onReportBatchTimeOut( System.Guid batchID )
    { /* Not implemented in this demo */ }

    public void onGetTearDownResult( ref bool[] resultREF )
    { /* Not implemented in this demo */ }

    public void onTearDownTimeOut()
    { /* Not implemented in this demo */ }

    // Private methods -----------------------------------------------------------
    private Measurement samplePhysicalMEM()
    {
        Measurement m = new Measurement();
        Process proc  = System.Diagnostics.Process.GetCurrentProcess();
        m.value       = proc.WorkingSet64.ToString();
        
        return m;
    }

    private Measurement sampleVirtualMEM()
    {
        Measurement m = new Measurement();
        Process proc  = System.Diagnostics.Process.GetCurrentProcess();
        m.value       = proc.VirtualMemorySize64.ToString();

        return m;
    }
}

} // namespace